design/lakectl-testing.md

```diff
@@ -0,0 +1,102 @@
+
+## Requirements
+* Usable both locally and as part of CI/CD
+* Ease of use - adding new tests should be as easy as possible. Moreover, adding new tests, or modifying existing ones, should not draw modifications to the testing framework itself
+  * It is obvious that changing the test **format** (e.g. expanding) will require framework modifications, in order to support the modified format
+* Decoupling of Framework and Data - the test data format should contain all the required info for the test itself, regardless of the framework design. In other words, we would like the ability to rewrite/replace the framework and still use the same tests data
+*
+
+## What We Want To Test 
+
+The goal is to the test `lakectl` correctness. At the very least we would like to verify that:
+* CLI input is parsed correctly
+* The correct requests are sent to the server (if any)
+* Server result is formatted correctly
+* Both erroneous input and error results from the server are formatted correctly
+* CLI Internal errors are handled correctly
+
+This testing framework is not designed to perform end-to-end tests (i.e. server correctness) or performance, for that matter
+
+## Architecture
+
+### Architecture Outline
+
+![lakectl testing architecture](./diagrams/lakectl-testing-arch.png)
+
+### Test Case Structure
+
+The test case structure should contain at least the following details:
+* pre-condition: the system state right before the tested command starts. This can be a function name to call, a mock type to use or a file containing lakectl commands to run in order to bring the system to that state (in case real server is in use)
+* lakectl command: a string with the `lakectl` command to test.
+* expected result: the expected output of a correct execution of the above command. This can be a string or, better, a file containing the output string
+
+In the future we might consider expending this structure with more fields, to allow extended functionality
+ 
+### How should we use `lakefs` backend
+
+Here are some of the pros and cons of each approach
+
+#### **Live Server**
+
+On the upside:
+* Straightforward - the implementation is already there, just bring the server up.
+  * Moreover, `lakefs` modifications are immediately reflected into the tests
+* The best possible simulation - this is the real server `lakectl` is communicating with
+
+And then, on the downside:
+* Bringing up the server requires additional dependencies - underlying storage, DB etc. While this supports the advantage of testing a **real** server environment, it complicates the test setup, and although this is easily solved by containers, it still worth mentioning
+* Performance-wise, creating a server environment for each test (thousands of them, ideally) is very expensive. We can, of course, group several tests (all of them?) to use a single server session, but this will eliminate the isolative and stateless nature of the tests, which is undesired
```

Comment:
_Commented by @itaiad200_
Yeah, true isolation sucks. The way we achieve isolation in Nessie is by creating a different repo for each test-case. It's not 100% isolated like running a fresh server, but it's a decent tradeoff.

Reply:
Using different repos is an option, given we decide to go we `lakefs` live server. Or, we can decide on mocking `lakefs` and get 100% isolation plus mocking improbable server states